network-api/networkapi/minisites/models.py

```diff
@@ -0,0 +1,363 @@
+            ('btn-secondary', 'Secondary button'),
+            ('btn-success', 'Success button'),
+            ('btn-info', 'Info button'),
+            ('btn-warning', 'Warning button'),
+            ('btn-error', 'Error button'),
+            ('btn-ghost', 'Ghost button'),
+        ],
+        default='btn-info',
+    )
+
+    outline = blocks.BooleanBlock(
+        default=False,
+        required=False
+    )
+
+    class Meta:
+        icon = 'link'
+        template = 'minisites/blocks/link_button_block.html'
+        value_class = LinkButtonValue
+
+
+class ImageTextBlock(blocks.StructBlock):
+    text = blocks.CharBlock()
+    image = ImageChooserBlock()
+    ordering = blocks.ChoiceBlock(
+        choices=[
+            ('left', 'Image on the left'),
+            ('right', 'Image on the right'),
+        ],
+        default='left',
+    )
+
+    class Meta:
+        icon = 'doc-full'
+        template = 'minisites/blocks/image_text_block.html'
+
+
+class VerticalSpacerBlock(blocks.StructBlock):
+    rem = blocks.IntegerBlock()
+
+    class Meta:
+        icon = 'arrows-up-down'
+        template = 'minisites/blocks/vertical_spacer_block.html'
+        help_text = 'the number of "rem" worth of vertical spacing'
+
+
+class VideoBlock(blocks.StructBlock):
+    url = blocks.CharBlock(
+        help_text='Please make sure this is a proper embed URL, or your video will not show up on the page.'
+    )
+    width = blocks.IntegerBlock(default=800)
+    height = blocks.IntegerBlock(default=450)
+
+    class Meta:
+        template = 'minisites/blocks/video_block.html'
+
+
+class iFrameBlock(blocks.StructBlock):
+    url = blocks.CharBlock(
+        help_text='Please note that only URLs from white-listed domains will work.'
+    )
+    width = blocks.IntegerBlock(default=800)
+    height = blocks.IntegerBlock(default=450)
+
+    class Meta:
+        template = 'minisites/blocks/iframe_block.html'
+
+
+"""
+We'll need to figure out which components are truly "base" and
+which are bits that should be used in subclassing template-based
+page types.
+"""
+base_fields = [
+    ('heading', blocks.CharBlock()),
+    ('paragraph', blocks.RichTextBlock()),
+    ('image_text', ImageTextBlock()),
+    ('image', ImageChooserBlock()),
+    ('video', VideoBlock()),
+    ('iframe', iFrameBlock()),
+    ('linkbutton', LinkButtonBlock()),
+    ('spacer', VerticalSpacerBlock()),
+]
+
+
+class ModularPage(Page):
+    """
+    The base class offers universal component picking
+    """
+
+    header = models.CharField(
+        max_length=250,
+        blank=True
+    )
+
+    body = StreamField(base_fields)
+
+    content_panels = Page.content_panels + [
+        FieldPanel('header'),
+        StreamFieldPanel('body'),
+    ]
+
+    show_in_menus_default = True
+
+
+class MiniSiteNameSpace(ModularPage):
+    subpage_types = [
+        'CampaignPage',
+        'OpportunityPage',
+    ]
+
+    """
+    This is basically an abstract page type for setting up
```

Comment:
Should this model be specifically defined as abstract? https://docs.djangoproject.com/en/2.0/topics/db/models/#abstract-base-classes

Reply:
no, because it's "basically" an abstract page, except it gets used to make concrete pages for the CMS to have top level "opportunity" / "campaign" / (whatever) roots.